fish.go

```diff
@@ -158,20 +159,17 @@ func (a *App) prepareFishFlags(flags []Flag, previousCommands []string) []string
 }
 
 func fishAddFileFlag(flag Flag, completion *strings.Builder) {
-	switch f := flag.(type) {
-	case *GenericFlag:
-		if f.TakesFile {
-			return
-		}
-	case *StringFlag:
-		if f.TakesFile {
-			return
-		}
-	case *StringSliceFlag:
-		if f.TakesFile {
+	val := reflect.ValueOf(flag)
+	// if flag is a non-nil pointer to a struct...
+	if val.Kind() != reflect.Invalid && val.Elem().Kind() == reflect.Struct {
```

Comment:
It's not safe to call `val.Elem()` unless we know that the type has an element. 

From the Godoc:
```go
// Elem returns the value that the interface v contains
// or that the pointer v points to.
// It panics if v's Kind is not Interface or Ptr.
// It returns the zero Value if v is nil.
```

And an example:

```go
// Basic version of this function for demonstration
func mightPanic(flag interface{}) {
	val := reflect.ValueOf(flag)
	if val.Kind() != reflect.Invalid && val.Elem().Kind() == reflect.Struct {
		field := val.Elem().FieldByName("TakesFile")
		if field.Kind() == reflect.Bool && field.Bool() {
			return
		}
	}

}


func main() {
	type impl struct {
		TakesFile bool
	}
	inst := impl{TakesFile: true}
	var empty *impl
	mightPanic(&inst)             // safe
	mightPanic(empty)             // safe
	mightPanic(map[string]bool{}) // panic
	mightPanic(inst)              // panic
	mightPanic(5)                 // panic
	mightPanic("oh no")           // panic
}
```

I believe the fix is as simple as swaping in `== reflect.Ptr` for `!= reflect.Invalid`—we need the value to be `reflect.Ptr` and the element to be `reflect.Struct`, and those both mean they are not invalid. 

Still, I have a few concerns with the implementation. 

The first is that the test cases did not catch the fact that any non-pointer implementation of `Flag` would cause a panic. The other is that while we're likely to catch additional internal implementations in the future, this is not a generic implementation, it's a check to see if the flag implementation happens to be a pointer to a struct that has a particular field with a particular name and type. The good news is it's unlikely that there are any implementations for which that isn't a safe assumption, and I do think that's more of a flaw with the current design rather than with this change.

I think there's two options going forward.
1. Revert back to the first commit in this MR that adds the `PathFlag` case, which we should be pretty safe to assume works, because it's effectively extending what we were already doing
2. Keep this pattern, but write sufficient test coverage such that we can be confident that for any implementation of a flag, this produces the correct output without panicking.

Reply:
Great note, will do :)

Reply:
I pushed 279c84210adff5a98f53d221d72f57f9d16b9b67 to handle unusual types, but as I began expanding the tests to exercise more possible inputs I concluded that the reflection-based approach really just introduces more edge cases than it resolves. It might be worth revisiting if the problem comes up again with another Flag implementation, but for now I've reverted this branch to the switch/case version.